### ISSUE TITLE
Multiple Activity annotated as AndroidEntryPoint in different modules with Hilt not working

### ISSUE THREAD

**Issue by rhlmshr** [google/dagger#2191]
I have two modules with separate activity in each of them.

**(:app) Module**
- MainApplication
```
@HiltAndroidApp
class MainApplication : Application() 
```

- MainActivity
```
@AndroidEntryPoint
class MainActivity : AppCompatActivity(R.layout.activity_main) {    
    fun onClickMeClicked(view: View) {
        startActivity(Intent(this, SecondActivity::class.java))
    }
}
```

**("dummyFeature") Module**
```
@AndroidEntryPoint
class SecondActivity : AppCompatActivity(R.layout.activity_second)
```

As soon as I am annotating the SecondActivity with AndroidEntryPoint, and invoking the onClick function from MainActivity it is throwing below mentioned Runtime exception:

> E/AndroidRuntime: FATAL EXCEPTION: main
>     Process: com.curieo.multimodulehilt, PID: 18269
>     java.lang.RuntimeException: Unable to start activity ComponentInfo{com.curieo.multimodulehilt/com.curieo.dummyfeature.SecondActivity}:
> java.lang.ClassCastException:
> com.curieo.multimodulehilt.DaggerMainApplication_HiltComponents_SingletonC$ActivityRetainedCImpl$ActivityCImpl
> cannot be cast to
> com.curieo.dummyfeature.SecondActivity_GeneratedInjector
>         at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3449)
>         at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3601)
>         at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
>         at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
>         at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
>         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
>         at android.os.Handler.dispatchMessage(Handler.java:106)
>         at android.os.Looper.loop(Looper.java:223)
>         at android.app.ActivityThread.main(ActivityThread.java:7656)
>         at java.lang.reflect.Method.invoke(Native Method)
>         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
>      Caused by: java.lang.ClassCastException: com.curieo.multimodulehilt.DaggerMainApplication_HiltComponents_SingletonC$ActivityRetainedCImpl$ActivityCImpl
> cannot be cast to
> com.curieo.dummyfeature.SecondActivity_GeneratedInjector
>         at com.curieo.dummyfeature.Hilt_SecondActivity.inject(Hilt_SecondActivity.java:62)
>         at com.curieo.dummyfeature.Hilt_SecondActivity.onCreate(Hilt_SecondActivity.java:36)
>         at android.app.Activity.performCreate(Activity.java:8000)
>         at android.app.Activity.performCreate(Activity.java:7984)
>         at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309)
>         at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3422)
>         at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3601) 
>         at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85) 
>         at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
>         at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
>         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066) 
>         at android.os.Handler.dispatchMessage(Handler.java:106) 
>         at android.os.Looper.loop(Looper.java:223) 
>         at android.app.ActivityThread.main(ActivityThread.java:7656) 
>         at java.lang.reflect.Method.invoke(Native Method) 
>         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592) 
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947) 

@bcorso

**Comment by bcorso** [comment]
This is probably a similar issue to #2123 and #1991.

This will happen if you have dependencies like `:app => :library => :dummyFeature`. We're currently looking at ways to fix this, and you can follow #1991 for that.

The work around for now is typically to either make `:app => :library` an `api` dependency so that it has access to the transitive `:dummyFeature` classes, or make `:app` depend directly on `:dummyFeature`.

**Comment by bcorso** [comment]
Closing this as dupe of #2123 and #1991.
